### PR TITLE
kata_agent: Enable devicemapper support for block devices

### LINKED ISSUE #622: rootfs failure observed with devmapper when testing in kata-runtime
When revendoring to latest virtcontainers and using docker with devicemapper as the storage driver, the following error is observed:
Feb 14 09:05:06 eernstworkstation kata-runtime-cc[42880]: time="2018-02-14T09:05:06-08:00" level=error msg="rpc error: code = Unknown desc = Could not run process: container_linux.go:296: starting container process caused \"exec: \\\"sh\\\": executable file not found in $PATH\"" source

Revendor pulls in the following changes:
```
98ccd53 (HEAD -> master, origin/master, origin/HEAD) Merge pull request #619 from containers/update_dep_vendoring
97be8e1 (origin/update_dep_vendoring) vendor: Update Gopkg.toml for new prune syntax
a0ee7ef vendor: Remove constraints from packages not directly vendored
27e2964 Merge pull request #618 from miaoyq/move-cc-shim-to-pod-netns
195d71f Run cc-shim in the network namespace of pod
6621b5d Merge pull request #617 from containers/fix_crio_race
62e5145 container: Let some time before to force kill a container process
8633d51 Merge pull request #609 from containers/improve_code
9ab4716 container: Remove useless container fetching
1f9f41f pod: Remove useless pod fetching
b19559f pod: Don't update container states from the pod
f8d2c11 Merge pull request #528 from jcvenegas/hypervisor-netowrking-model
e047779 net: remove netModel string function
ac880a1 test: network: add unit test
f19bd7a oci: Allow to get InternetworkingModel from config
573dcf0 network: Add NetInterworkingModel methods
6d19f56 network: Endpoint type dictate how the connection needs to be made
037a81a network: define internetworking model at endpoint creation
089119a network: remove unsed function createNetworkEndpoints
254ad2a Merge pull request #591 from devimc/network/io
5a1f1bb Merge pull request #607 from miaoyq/dockershim-support
e6c6f84 Add support for dockershim pod annotations
3ecda86 Merge pull request #593 from containers/fix_mount
e01404c container: Allow to send a signal to a created container
e5ad775 container: Allow to stop a container in state created
00a9091 qemu: change CPU topology to support until 240 CPUs
7dc7483 mount: Explicitely modify and store container mount list
9acfae9 container: Move devices and mounts creation to createContainer()
9a81686 Merge pull request #592 from containers/fix_container_semantic
484cc19 (origin/fix_container_semantic) api: Add new container pointer to pod structure
d07aeb2 container: Replace fetchContainer() with findContainer()
4f82905 Merge pull request #573 from jodh-intel/safe-clean-rule
80038f3 build: Only clean files if they exist
a4377a7 build: Fix clean and uninstall
```

The error is introduced with patch:

```
commit 9acfae94caa8800136d151f3edac0e8292f21949
Author: Sebastien Boeuf <sebastien.boeuf@intel.com>
Date:   Thu Feb 1 17:34:54 2018 -0800

    container: Move devices and mounts creation to createContainer()
    
    Instead of waiting for the start a container to get its mounts and
    devices processed, this commit moves those actions earlier in the
    sequence, that is to CreateContainer().
    
    The reason behind this is that in case of Kata Containers, we expect
    all resources to be created so that we can actually create the
    container from the agent, when CreateContainer() is called.
    
    Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

```

### PR DESCRIPTION
**Pull request by sboeuf**
This PR enables the support for Kata with block devices. It enables it both for rootfs being passed as a block device from devicemapper, and for block devices being passed through the VM to the container.

Fixes #622

### CODE REVIEW

**Comment by sboeuf** [comment]
Depends on https://github.com/kata-containers/agent/pull/133

**Comment by sboeuf** [comment]
cc @sameo @jodh-intel

**Comment by sameo** [comment]
LGTM

**Comment by sameo** [comment]
@sboeuf We should probably wait until kata-containers/agent#133 gets merged before taking that one.

**Comment by sboeuf** [comment]
@sameo yes I agree since it depends on it, and I don't want to merge this because the agent PR might change based on comments.
I have added `do-not-merge` label just to make sure we don't forget !

File: kata_agent.go [hunk 1]
```diff
@@ -474,9 +476,13 @@ func (k *kataAgent) createContainer(pod *Pod, c *Container) (*Process, error) {
 		}
 
 		rootfs.Source = filepath.Join(devPath, driveName)
-		rootfs.MountPoint = rootPath // Should we remove the "rootfs" suffix?
+		rootfs.MountPoint = rootPathParent
 		rootfs.Fstype = c.state.Fstype
 
+		if c.state.Fstype == "xfs" {
```
Comment (jodh-intel) [inline kata_agent.go]: Do we have any XFS tests in the `tests` repo to exercise this? If not, maybe you could add a small unit test?

File: kata_agent.go [hunk 2]
```diff
@@ -42,12 +42,13 @@ var (
 	defaultKataID               = "charch0"
 	errorMissingProxy           = errors.New("Missing proxy pointer")
 	errorMissingOCISpec         = errors.New("Missing OCI specification")
-	kataHostSharedDir           = "/tmp/kata-containers/shared/pods/"
-	kataGuestSharedDir          = "/tmp/kata-containers/shared/pods/"
+	kataHostSharedDir           = "/run/kata-containers/shared/containers/"
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply kata_agent.go]: Done

File: kata_agent.go (hunk 1 above)
Comment (sboeuf) [reply kata_agent.go]: How can I add a small unit test for that ? I would have to test `createContainer()`, which is a very hard one...

Comment (jodh-intel) [reply kata_agent.go]: I was thinking you could create an invalid `pkg.oci.config` annotation. Then, when you call `createContainer(pod *Pod, c *Container)`, the `json.Unmarshal()` will fail. But `c.state.Fstype`  should then be `"xfs"`.

File: kata_agent.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (amshinde) [reply kata_agent.go]: In the commit message you mentioned guest path is being updated to include "containers", but you are updating the host path here.

[the remark this answers is not in the document]
Comment (sboeuf) [reply kata_agent.go]: Oh yes you're right, I have done the opposite of what I had written !

**Comment by sboeuf** [comment]
@amshinde PR updated, PTAL

File: kata_agent.go (hunk 1 above)
Comment (jodh-intel) [reply kata_agent.go]: Looking again, I agree - you can't test for this scenario as even if you can force the error, the root mount options aren't returned.

**Comment by sboeuf** [comment]
I have checked this PR with latest version of https://github.com/kata-containers/agent/pull/133, and all cases works as expected:
- Overlay rootfs with no device passthrough
- Overlay rootfs with block device passthrough
- Devicemapper rootfs with no device passthrough
- Devicemapper rootfs with block device passthrough

We'll need a follow up PR to be able to support block devices with SCSI, but we should move forward and get this one merged first.

**Comment by sboeuf** [comment]
https://github.com/kata-containers/agent/pull/133 has been merged.
@grahamwhaley @jodh-intel please merge this if the CI is all green

**Pull request merged**